### PR TITLE
Addition of 'App Instance ID' and 'Core Instance ID'

### PR DESCRIPTION
**Pull request by shoamano83**
This document proposes to include two IDs in the payload of Start Service and Start Service ACK frames during Version Negotiation. This is a sub-proposal of SDL-nnnn: Supporting simultaneous multiple transports.

### CODE REVIEW

**Comment by shoamano83** [comment]
Hi @theresalech, this is ready for review. Thanks.

File: proposals/nnnn-mt-instance-ids.md [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+  - Advantage: this approach will not require high-quality random generator
+  - Disadvantage: this approach requires three things compared to Version 4:
+    1. current time. Concern is that some embedded systems may take some time to acquire current time after ignition on (e.g. to receive GPS time).
+    2. MAC address, either Wi-Fi or Bluetooth
+    3. file system to store clock sequence and last timestamp
+* Using UUID Version 5 (name-based UUID) for Core Instance ID and employ ISO VIN code as the names.
+  - Advantage: this approach generates truly unique IDs (provided that there is no duplicate VINs on earth)
+  - Disadvantages: Core needs to acquire VIN from HMI before starting transports.<br>
+    There may be a privacy concern as generated UUID has no randomness. An attacker might be able to obtain VIN from Core Instance ID by calculating hashes of all possible VINs offline.
+* Exchanging the IDs through RegisterAppInterface request and response instead of Version Negotiation. As noted in [SDL-nnnn][multiple_transports], this approach has more impacts on implementations since ApplicationManager layers of SDL Core and Proxies need to be updated.
```
Comment (theresalech) [inline proposals/nnnn-mt-instance-ids.md]: Please update to "impact"

File: proposals/nnnn-mt-instance-ids.md [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+* Instead of Proxy generating App Instance ID, Core generates it and notify it to Proxy when the first transport is connected. The advantage of this idea is that Core can regenerate App Instance ID if it collides with previously generated one. However, Core will consume more and more RAM as it needs to remember all IDs that have been generated.
```
Comment (theresalech) [inline proposals/nnnn-mt-instance-ids.md]: Please update to "...Core generates it and notifies it to Proxy when..."

File: proposals/nnnn-mt-instance-ids.md [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+- ICCID (IC Card Identifier) and IMSI (International Mobile Subscriber Identity) - Not available on iPhone.
+- UDID (Unique Device Identifier) - iOS specific. An app that accesses this ID is rejected by Apple.
+- identifierForVendor property of UIDevice - iOS specific. The issue with this ID is that it is shared between apps from same vendor. We need to combine it with some kinds of app's ID (such as bundle ID) to make it application-specific.
+- advertisingIdentifier property of ASIdentifierManager - iOS Specific. This is not suitable for our purpose as it is intended for advertisements. Also note that the phone's user can disable this feature.
+- Android ID - Android specific. This is still available, however [an entry of Android developer's blog][androido] suggests that *"Developers requiring a device-scoped identifier, should instead use a resettable identifier, such as Advertising ID, giving users more control."*
+- Advertising ID - Android specific. This is not suitable for our purpose as it is primary intended for advertisements. Also, this introduces a dependency on Google Play Services.
```
Comment (theresalech) [inline proposals/nnnn-mt-instance-ids.md]: Please update to "...as it is primarily intended for advertisements."

File: proposals/nnnn-mt-instance-ids.md [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+* Impacted Platforms: [Core / iOS / Android / Protocol / BSON library]
+
+## Introduction
+
+This document proposes to include two IDs in the payload of Start Service and Start Service ACK frames during Version Negotiation. This is a sub-proposal of [SDL-nnnn: Supporting simultaneous multiple transports][multiple_transports].
+
+
+## Motivation
+
+To support multiple transports between Core and Proxy, Core needs to distinguish between the cases "a single SDL app connecting to SDL Core using two different transports" and "same SDL apps on two phones connecting to SDL Core using different transports". Also, Proxy needs to make sure that it is not connected to different head units at the same time.
```
[the remark this answers is not in the document]
Comment (shoamano83) [reply proposals/nnnn-mt-instance-ids.md]: I have slightly updated the sentence for better clarification. Please let me know if it makes sense to you.

**Comment by shoamano83** [comment]
Thanks @theresalech, I updated the document to address your comments.

**Pull request merged**